File: Recursion/344. Reverse String.py
class Solution(object):
    def reverseString(self, s):
        """
        :type s: List[str]
        :rtype: None Do not return anything, modify s in-place instead.
        """
        def helper(s, left, right):
            if left > right:
                return
            s[left], s[right] = s[right], s[left]
            left += 1
            right -= 1
            helper(s, left, right)
        helper(s, 0, len(s)-1)
        return
        

File: Recursion/344. Reverse String.cpp
/*
344. Reverse String
Two Pointers, String

Description:
Write a function that reverses a string. The input string is given as an array of characters s.

Example 1:
Input: s = ["h","e","l","l","o"]
Output: ["o","l","l","e","h"]

Example 2:
Input: s = ["H","a","n","n","a","h"]
Output: ["h","a","n","n","a","H"]

Constraints:
1 <= s.length <= 105
s[i] is a printable ascii character.
*/

// This is a solution using recursive approach.
class Solution {
public:
    void reverseString(vector<char>& s) {
        helper(s, 0, s.size()-1);
        return;
    }
    void helper(vector<char>& str, int left, int right){
        if(left>right){
            return;
        }
        swap(str[left], str[right]);
        helper(str, ++left, --right);
    }
};
